design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+    # aliases of the primary fqdn
+    aliases:
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # are described in fqdn and aliases, the tls.secretName secret must contain a
+    # matching certificate
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # service defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    service:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
+  - match: /ads
+    # more than one service name/port pair may be present allowing the services for a route to
+    # be handled by more than one service.
+    service:
+    - name: ads-red
+      port: 9090
+    - name: ads-blue
+      port: 9090
+```
+
+This is an example of the `google-finance` object which has been delegated responsibilty for paths starting with `/finance`.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google-finance
+  namespace: finance
+spec:
+  # note that this is an edge, so there is no virtualhost key
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific, however as this is an edge, only prefixes that match the prefix
+  # that delegated to this object.
+  routes:
+  # each route in this object must start with /finance as this was the prefix which
+  # delegated to this object.
+  - match: /finance
+    service:
+    - name: finance-app
+      port: 9999
+  - match: /finance/static
+    service:
+    - name: finance-static-content
+      port: 8080
+  - match: /finance/partners
+    # delegate the /finance/partners prefix to the partners namespace
+    delegate:
+      name: finance-partners
+      namespace: partners
+```
+
+## Delegation rules
+
+The delegation rules applied are as follows
+
+1. If an `IngressRoute` object contains a `spec.virtualhost` key it is considered a root.
```

Comment:
@jbeda  was pretty sold on the polymorphic nature of reusing one object, but now you bring it up it could make validation simpler; IngressRoute and, say, DelegateRoute, which does not contain a virtualhost stanza at all would be more obvious.

My aversion to the designs we've talked about previously with a VirtualHostCRD and RouteCRD is they don't scale down to the simple use case of, "i'm in my own cluster, I want to host my own web application, I'm not looking for multitennancy protections".

With that said, having a IngressRoute and DelegateRoute doesn't solve the problem of being construct a delegation path from a virtualhost that you control, to a set of routes that someone else controls. I'm not opposed to having two CRDs, and that may be easier in cases where you want to provide read only access to the IngressRoute CRD without necessarily letting it be editable.

Reply:
The previous model would work because we had a `Route` and a `Policy`, the policy is optional, but not relevant to this design. 

One thing missing is the `type` field from the original design doc: ` “type” field that is either “Root” or “Delegated”`

I like splitting CRD's, but maybe we can utilize the type field and gain the same.

Reply:
@jbeda i'm not grossed out by IngressRoute and DelegateRoute (to pull names out of hat) having a lot of crossover. At the YAML level and the gRPC level and the Go types level this has little overhead.

What having two objects would give us is

- the simple, KUARD, case, remains a single object. We don't have to do the VHost CRD and Route CRD for the simple case
- having two objects would give an option for RBAC to be applied according to site specific requirements. It would mean users of the /finance namespace could have permission to _view_ the IngressRoute root CRD which contains the data about their route, but this object would in tern delegate to a DelegateRoute in the same namespace which users could edit. This wouldn't solve the problems of handling TLS secrets in the same namespace (RBAC isn't much help unless you want to prohibit normal users of that namespace from using ANY secret objects) or the issue where a single VHost is delegated across multiple namespaces.

I think there are some attractive properties to warrant pursuing this approach. I'm also strongly influenced by notions of namespaced and non namespaced objects. IngressRoute could, perhaps, be sans namespace, with DelegateRoute could be namespaced.